lightning/src/ln/channelmanager.rs

```diff
@@ -3321,6 +3377,100 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			self.finish_force_close_channel(failure);
 		}
 	}
+
+	fn set_payment_hash_secret_map(&self, payment_hash: PaymentHash, payment_preimage: Option<PaymentPreimage>, min_value_msat: Option<u64>, invoice_expiry_delta_secs: u32, user_payment_id: u64) -> Result<PaymentSecret, APIError> {
+		assert!(invoice_expiry_delta_secs <= 60*60*24*365); // Sadly bitcoin timestamps are u32s, so panic before 2106
+
+		let payment_secret = PaymentSecret(self.keys_manager.get_secure_random_bytes());
+
+		let _persistence_guard = PersistenceNotifierGuard::new(&self.total_consistency_lock, &self.persistence_notifier);
+		let mut payment_secrets = self.pending_inbound_payments.lock().unwrap();
+		match payment_secrets.entry(payment_hash) {
+			hash_map::Entry::Vacant(e) => {
+				e.insert(PendingInboundPayment {
+					payment_secret, min_value_msat, user_payment_id, payment_preimage,
+					// We assume that last_node_announcement_serial is pretty close to the current
+					// time - its updated when we receive a new block with the maximum time we've
+					// seen in a header, and only incremented by one when we create a node
+					// announcement. As long as we aren't creating a new node announcement more
+					// often than once per second, it should never be more than two hours in the
+					// future. Thus, we add two hours here as a buffer to ensure we absolutely
+					// never fail a payment too early.
```

Comment:
Expiration is gateway on header time superior to old_serial so the risk is more never expiring a never-received inbound payment ?

Also I don't think the relation between "once per second" and "more than two hours in the future" is really clear. The problem is last_announcement_serial getting forever in advance of the block time, i.e if you generate more than 600 updates per-expected block interval. Of course your header time can still jump to 2 hours in the future per-consensus rules and swallow back the delay...and then your last_announcement_serial jumping ahead in the time race again. Not clearly easy to explain and reasoning on.

Reply:
Hmm, do you have a concrete suggestion to improve it? I'll go ahead and drop the reliance on `last_node_announcement_serial` here and just add a new field to track highest-header, which simplifies it some, but I'm not sure how to further simplify it.